### PR TITLE
Fix: sort document reference by long type id

### PR DESCRIPTION
**Pull request by milaGGL**
Document ID supports strings and well as long integers in the format of ` "__id<Long>__"` .  When sorting documents by document ID, it should be sorted in the following order:
1. Long (numeric order)
2. String (lexicographic order)

### CODE REVIEW

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java [hunk 1]
```diff
@@ -644,6 +647,98 @@ public void shutdownNowPreventsAddingNewListener() throws Exception {
...
+
+    firestore
+        .batch()
+        .set(col.document("A"), Collections.singletonMap("a", 1))
+        .set(col.document("a"), Collections.singletonMap("a", 1))
+        .set(col.document("Aa"), Collections.singletonMap("a", 1))
+        .set(col.document("7"), Collections.singletonMap("a", 1))
+        .set(col.document("12"), Collections.singletonMap("a", 1))
+        .set(col.document("__id7__"), Collections.singletonMap("a", 1))
+        .set(col.document("__id12__"), Collections.singletonMap("a", 1))
```
Comment (TomasMorton) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:662]: You might also consider test cases such as `_id1__` and `__id1_`, which are strings.

**Review by ehsannas — COMMENTED** [review]
LGTM. A minor improvement suggested below

**Review by ehsannas — APPROVED** [review]

**Comment by ehsannas** [comment]
We should add a testcase for negative numeric ID if that's allowed.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java [hunk 2]
```diff
@@ -644,6 +647,103 @@ public void shutdownNowPreventsAddingNewListener() throws Exception {
     listenerAssertions.hasError();
   }
 
+  @Test
+  public void snapshotListenerSortsQueryByDocumentIdInTheSameOrderAsServer() throws Exception {
```
Comment (milaGGL) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:651]: add negative number id

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BasePath.java [hunk 3]
```diff
@@ -123,16 +127,42 @@ boolean isPrefixOf(BasePath<B> path) {
...
+    } else if (!isLhsNumeric && isRhsNumeric) { // Only rhs is numeric
+      return 1;
+    } else if (isLhsNumeric && isRhsNumeric) { // both numeric
+      return Long.compare(extractNumericId(lhs), extractNumericId(rhs));
+    } else { // both non-numeric
+      return lhs.compareTo(rhs);
+    }
+  }
+
+  // Checks if a segment is a numeric ID (starts with "__id" and ends with "__").
```
Comment (TomasMorton) [inline google-cloud-firestore/src/main/java/com/google/cloud/firestore/BasePath.java:156]: Nit: Is there a reason not to write these method comment as a javadoc? `/** ... */`?

**Review by TomasMorton — APPROVED** [review]

**Pull request merged**